cloud/gcp: add custom retryer for gcs storage, retry on stream INTERNAL_ERROR

rhu713 opened the pull request:
Currently, errors like
`stream error: stream ID <x>; INTERNAL_ERROR; received from peer`
are not being retried. Create a custom retryer to retry these errors as
suggested by:

https://github.com/googleapis/google-cloud-go/issues/3735
https://github.com/googleapis/google-cloud-go/issues/784

Fixes: #85217, #85216, #85204, #84162

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/85024)
<!-- Reviewable:end -->

Commit:
cloud/gcp: add custom retryer for gcs storage, retry on stream INTERNAL_ERROR

Currently, errors like
`stream error: stream ID <x>; INTERNAL_ERROR; received from peer`
are not being retried. Create a custom retryer to retry these errors as
suggested by:

https://github.com/googleapis/google-cloud-go/issues/3735
https://github.com/googleapis/google-cloud-go/issues/784

Fixes: #85217, #85216, #85204, #84162

Release note: None

adityamaru commented on pkg/cloud/gcp/gcs_storage.go:
```diff
@@ -343,6 +345,24 @@ func (g *gcsStorage) Close() error {
+		}
+	}
+
+	if e := (errors.Wrapper)(nil); errors.As(err, &e) {
```
This looks correct but is it more intuitive to write `if ok := e.(errors.Wrapper); ok { ... }`? @knz is there a more canonical way to check if an error is wrapped?

rhu713 replied:
FYI I had to write it like this because of a checker that we have. This pattern is the suggested one:
https://github.com/cockroachdb/cockroach/blob/master/pkg/testutils/lint/passes/errcmp/errcmp.go#L97

adityamaru replied:
OH! never mind me then 👍

adityamaru reviewed (approved):
Nice, this'll solve a bunch of failing roachtests 🤞 LGTM, it's a bummer they haven't exposed `ShouldRetry` yet but keeping the logic in a different file is a good idea.

adityamaru commented:
Before merging can you also add all the failing roachtests on master as `Fixes` so they auto-close?

msbutler commented on pkg/cloud/gcp/gcs_storage.go:
```diff
@@ -343,6 +345,24 @@ func (g *gcsStorage) Close() error {
+	if defaultShouldRetry(err) {
+		return true
+	}
+
```
maybe add a small comment on these specific cases are outside google's default retries, and why we retry them?

rhu713 replied:
done

rhu713 commented:
bors r+

rhu713 commented:
blathers backport 22.1
